### ISSUE TITLE
Proposal: File header validation analyzer and fix

### ISSUE THREAD

**Issue by sharwell** [dotnet/roslyn#33012]
## Summary

Currently the IDE does not offer built-in support for configuring headers (aka file banners or copyright headers) expected to appear in source files. This proposal defines a simple analyzer and code fix combination to support a majority of cases used in practice.

When configured, Roslyn would be responsible for adding a file header to newly-created source files it recognizes. Other language services would have access to the **.editorconfig** configuration and could follow suit (the property is defined in a manner that others could follow).

## Variables

The following built-in variables may be used in defining the expected file header content.

| Variable | Value | 
| --- | --- |
| `fileName` | The name of the source file, including its extension |

## Configuration

The expected header content is provided as a template. Substitution variables may be referenced in curly braces, such as `{fileName}`. The sequence `\n` may be used for a single newline in the document.

| Property | Meaning |
| --- | --- |
| `file_header_template` | The template text for file headers. |

## Header forms

File headers in C# source files are allowed to have any of the following forms. The first form is preferred, meaning it is used when the code fix inserts a new comment into a file which did not previously have one.

```csharp
// line 1
// line 2
// line 3
// ...
```

```csharp
/* line 1
 * line 2
 * line 3
 * ...
 */
```

```csharp
/* line 1
   line 2
   line 3
   ... */
```

### Single-line header

The following configuration could be used for the code in this repository prior to #25423:

```
file_header_template = Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
```

Files would be expected to start with the comment:

```csharp
// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
```

### Multi-line header

The following could be used to implement the three-line header requested in #25423:

```
file_header_template = Licensed to the .NET Foundation under one or more agreements.\nThe .NET Foundation licenses this file to you under the MIT license.\nSee the LICENSE file in the project root for more information.
```

Files would be expected to start with the comment:

```csharp
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.
```

### StyleCop-compatible header

The following could be used to define a simple StyleCop-compatible XML header:

```
file_header_template = <copyright file="{fileName}" company="Your Company Here">\nYour copyright text here\n</copyright>
```

The file **MyFile.cs** would be expected to start with the comment:

```csharp
// <copyright file="MyFile.cs" company="Your Company Here">
// Your copyright text here
// </copyright>
```

### What about a `{currentYear}` variable?

**tl;dr:** The `currentYear` variable is not required. Simply replace `{currentYear}` in your template with the year you expect to appear in file headers.

Implementing validation using the year as a variable is highly problematic in practice for two reasons:

1. The success of the build depends on when the build runs. It is impossible to create commits which will build successfully a year later.
2. Forcing the header to a later date via analysis would result in a post-dated copyright notice for the code. Depending on the specific country, this is likely to result in an invalid copyright date, a situation which can result in the loss of certain types of copyright protection otherwise afforded to projects.

Fixed dates in the file header template support headers with the strongest copyright protections. Projects which require other approaches may not be able to use this analyzer, but please feel free to use the code for the built-in analyzer under its open source licensing terms to implement a package supporting other specific scenarios.

## Notable changes

The following are notable changes to this proposal as a result of comments in the thread below.

* Changed `dotnet_style_file_header_template` to `file_header_template` (per https://github.com/dotnet/roslyn/issues/33012#issuecomment-459870949)
* Clarify that Roslyn would automatically apply the header to new source files (per https://github.com/dotnet/roslyn/issues/33012#issuecomment-459870949)

**Comment by jasonmalinowski** [comment]
Overall, love the proposal, a few specific thoughts:

> dotnet_style_file_header_template

I wouldn't prefix it with "dotnet_style". First of all, it's not .NET specific. I would entirely expect that adding a new .props or .targets file would get an appropriate comment added. Or if I add C++ code. Yes, we will be enforcing this with a .NET analyzer, but that's just one of the places where this property should be consumed.

"style" seems odd anyways: this isn't a style question, it's a legal question -- the thing that beats all style questions in the end. :smile:

Did we investigate if there is already an .editorconfig convention for other languages/tooling that we should adopt instead of (potentially) doing NIH syndrome?

> What about a `{currentYear}` variable?

I agree, of course, with the legal justification. I do wonder though whether we could address the "commits that don't validate later" by changing how we would validate. Instead of having the analyzer require an _exact_ string, it'd instead take the template and basically convert it to a regex (or something) that simply filled in `{currentYear}` as "it should be a four digit year". And then confirm that the comment matches that.

I could imagine a similar request that people might want a "created by" header which sticks the username of the person creating the file; the validation for the regex becomes "there's a username-looking thing" there. And once again yes, that's what people _should_ use source control for, but reality often differs.

> Title: File header validation analyzer and fix

I assume we're also going to hook this up to the code that formats newly create files from templates so we add this header to files immediately upon adding new files to your project. The fix is obviously there to apply/update existing files and projects, but if I add a new file and the immediately have to invoke the fix, that's silly. Just wanted to make sure that was explicitly called out in the proposal. More generally, perhaps we should be running and applying _all_ code style analyzers, but I think this one would be the most immediately one.

**Comment by sharwell** [comment]
> Did we investigate if there is already an .editorconfig convention for other languages/tooling that we should adopt instead of (potentially) doing NIH syndrome?

There isn't anything I could find in [editorconfig/editorconfig](https://github.com/editorconfig/editorconfig).

> I do wonder though whether we could address the "commits that don't validate later" by changing how we would validate.

A `{createdYear}` variable would be possible to implement without the build breaking. When generating a file header, it would use the current year. Validation would pass for any year at or prior to the current year. However, it still runs the risk of post-dating source code as well as discrepancies between individual files and the root license notice for the software. I would prefer to omit this for now since it produces the safest and easiest overall experience.

Keep in mind that most validation solutions focus on _validation_, but this proposal covers more area. It is essential that the template define the header in a way that it can be added or corrected automatically in cases where validation fails. Generation of a correct header is much more difficult than validation when variables like times are involved.

> I could imagine a similar request that people might want a "created by" header which sticks the username of the person creating the file; the validation for the regex becomes "there's a username-looking thing" there.

This seems like the type of case where it would be reasonable to create a customized solution.

> I assume we're also going to hook this up to the code that formats newly create files from templates so we add this header to files immediately upon adding new files to your project. 

Yes, new files should apply the template automatically. This is similar to the requests in the following issues:

#27186
#22171

**Comment by jasonmalinowski** [comment]
As far as projects, saw this in some quick googling:

![image](https://user-images.githubusercontent.com/201340/52152169-4a631480-262a-11e9-89cc-2a30cd2b2079.png)

But wasn't sure how it was actually using it.

**Comment by jasonmalinowski** [comment]
It might also be worth just asking the .editorconfig community.

**Comment by sharwell** [comment]
> As far as projects, saw this in some quick googling:

That's a list of projects that use EditorConfig, i.e. have a .editorconfig in the repository. 😄 

> It might also be worth just asking the .editorconfig community.

I can create an issue there.

**Comment by jasonmalinowski** [comment]
> That's a list of projects that use EditorConfig, i.e. have a .editorconfig in the repository. 😄

D'oh. 🤦‍♂️

**Comment by sharwell** [comment]
Design review status: The IDE team is comfortable with the proposal above, provided the release documentation accompanying this feature cannot be misinterpreted as legal advice. Before implementation, we are looking for secondary reviews of the proposal from the VS Editor team and from the VS Code team to confirm that the proposal will be sufficient to support automation in other languages and/or environments.

**Comment by mhutch** [comment]
Squishing multiline headers one one line looks like it would be pretty hard to edit, especially with long headers e.g. the entire MIT license. Is there any precedent for multiline vaues in editorconfig files?

**Comment by sharwell** [comment]
@mhutch There is no support for multi-line values in .editorconfig files. Fortunately these values would not need to be frequently edited.

It might be possible to update the EditorConfig Language Service extension to include a designer dialog where multi-line values could be edited and then automatically escaped when the file is updated.

**Comment by jmarolf** [comment]
@sharwell @mhutch 

I would just have the value point to a relative path of a text file
```ini
file_header_template = ..\license.txt
```

And then just paste the entire contents of that text file into the comments

**Comment by JoeRobich** [comment]
> I would just have the value point to a relative path of a text file

@jmarolf I assume that is relative to the .editorconfig file and not the code files. Which would be a bit tricky since currently we don't get back information about which .editorconfig contributed which options.

**Comment by jmarolf** [comment]
>  I assume that is relative to the .editorconfig file and not the code files.

@JoeRobich that was how I envisioned it. and yes we would need to push through more data in the parser to resolve this. We should be able to do this anyways though. @mavasani how do we detect where to place options in editorconfigs today? Do we just try the first one we find?

**Comment by JoeRobich** [comment]
@jmarolf Currently, when changing the Severity for a diagnostic, we look through the AnalyzerConfigDocuments for the Project and choose based on longest matching file path. https://github.com/mavasani/roslyn/blob/e2f494603ca59c4f36cd4970f39712adc073f32a/src/Workspaces/Core/Portable/Shared/Extensions/ProjectExtensions.cs#L58-L102

**Comment by sharwell** [comment]
> I would just have the value point to a relative path of a text file

I don't think this is a particularly viable option (or at least doesn't outweigh the editing disadvantage of having the template in-line).

* Users don't want to add _yet another_ file to source control for configuration.
* This change would make it more difficult to unify with the rest of the .editorconfig ecosystem.
* Having the location of the .editorconfig is not sufficient, since multiple files can participate with inheritance. We would need the location of the _specific_ .editorconfig which defined the effective value for the property.

**Comment by CyrusNajmabadi** [comment]
Is there any reason we can't just put the desired copyright/template into .editorconfig itself? it's just a string right?

**Comment by jmarolf** [comment]
@CyrusNajmabadi interesting idea, we could put the sting in a comment in the editorconfig file

```ini
# [file_header_template_name]
# Licensed to the .NET Foundation under one or more agreements.
# The .NET Foundation licenses this file to you under the MIT license.
# See the LICENSE file in the project root for more information.

file_header_template = file_header_template_name
```

**Comment by canton7** [comment]
It might be worth looking at [License Header Manager](https://marketplace.visualstudio.com/items?itemName=StefanWenig.LicenseHeaderManager) to see what other license header functionality users are going to request. 

For example, the abilitity to insert different license headers into different file types (e.g. xaml, cshtml).

Is the plan to allow users to update the license header template and then update all files with the new license header?

**Comment by jmarolf** [comment]
> Users don't want to add yet another file to source control for configuration.

Thinking about this I do feel its sub-optimal. The compiler wants every input to be specified at the commandline. We don't want some arbitrary file effecting compilation that is not declared as part of the compilation

> This change would make it more difficult (likely impossible) to unify with the rest of the .editorconfig ecosystem.

@sharwell can you explain what you mean by this? How is the .editorconfig ecosystem blocked? I agree that whatever we do is going to stress the editorconfig format about as hard as the naming styles did. We should at least checkin with the community to ensure whatever format we decide on makes sense to others. 

> Having the location of the .editorconfig is not sufficient, since multiple files can participate with inheritance. We would need the location of the specific .editorconfig which defined the effective value for the property.

We own the parser for this so we can do what we want. I am sure we could find a way to write a parser that has the behavior we want. I don't want us to get bogged down with the exact implementation before we decide what the format should be.

**Comment by CyrusNajmabadi** [comment]
> For example, the abilitity to insert different license headers into different file types (e.g. xaml, cshtml).

Seems totally doable with .editorconfig.  it already has the concept of rules that are bucketed by file extension. I don't see why this would be any different.  @jmarolf do you concur?

**Comment by mavasani** [comment]
> we could put the sting in a comment in the editorconfig file

However, the comment strings are not available to analyzers via analyzer options. Are you suggesting that compiler should be updated to treat this option specially and read the value, and then parse the comment block with this value as a header and add a new key-value pair entry? This would be a compiler feature request.

> Thinking about this I do feel its sub-optimal. The compiler wants every input to be specified at the commandline. We don't want some arbitrary file effecting compilation that is not declared as part of the compilation

If we want any content to be in a separate file, it should be through an AdditionalFile with a special name and format that the analyzer understands.

Overall, I agree with @sharwell that we want to ideally have the following principle:
1. .editorconfig is the configuration mechanism for simple key-value pair options
2. Additional files is the configuration mechanism for more complex analyzer configuration inputs

I think we can write an analyzer with dual support - reads an editorconfig option for simple one line header and respects the special additional file for multi-line header.

Now if you are able to get the compiler feature request approved, then I am not against enhancing the editorconfig parsing to support multi-line values via comment blocks.

**Comment by CyrusNajmabadi** [comment]
>However, the comment strings are not available to analyzers via analyzer options. Are you suggesting that compiler should be updated to treat this option specially and read the value, and then parse the comment block with this value as a header and add a new key-value pair entry? This would be a compiler feature request.

Shouldn't the compiler provide *all* .editorconfig values to analyzers/fixers?

**Comment by CyrusNajmabadi** [comment]
> .editorconfig is the configuration mechanism for simple key-value pair options

This seems like a simple key-value pair option :D.  You have the key be something like `file_header_template`, and the value is a simple string containing the contents (maybe with some template pieces).

**Comment by mavasani** [comment]
Compiler's editorconfig parser skips all comment lines: http://sourceroslyn.io/#Microsoft.CodeAnalysis/CommandLine/AnalyzerConfig.cs,138.

> This seems like a simple key-value pair option

It is a cascaded option. If you query with key `file_header_template`, you will get value `file_header_template`. If you query with key `file_header_template`, you will get no entry. Compiler's editorconfig parser needs to understand a different format for multiline values.

**Comment by CyrusNajmabadi** [comment]
sorry, i'm dense... i don't know what you're saying.  it look like `file_header_template` in both cases.  can you clarify?

**Comment by CyrusNajmabadi** [comment]
Note: i'm not saying it shouldbe done as a comment.  I'm saying it should be done as a simple key with a string value.

**Comment by mavasani** [comment]
> Note: i'm not saying it shouldbe done as a comment. I'm saying it should be done as a simple key with a string value.

How will be support multiline header? .editorconfig parser treats each line as independent unit.

**Comment by CyrusNajmabadi** [comment]
I'd be 100% ok with:

```
file_header_template = Copyright 2016-2018, Pulumi Corporation.\n\nLicensed under the Apache License, Version 2.0 (the "License");
```

etc.

**Comment by mavasani** [comment]
@CyrusNajmabadi We are then actually on the same page. I think @sharwell's current proposal also takes that approach. I thought the latest discussion was centered around how to support https://github.com/dotnet/roslyn/issues/33012#issuecomment-513940616, i.e. allow multi-line headers without squishing them into a single line with newline characters.

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi We are then actually on the same page. I think @sharwell's current proposal also takes that approach.

Ah, sorry for the confusion then!

**Comment by mavasani** [comment]
I think we can still support https://github.com/dotnet/roslyn/issues/33012#issuecomment-513940616 by having the analyzer look for an additional file with specific name and format. Users can then decide if they prefer to encode their multiline header inside the .editorconfig file with newline characters within a single line OR have it as multiline text in a separate additional file.

**Comment by jmarolf** [comment]
makes sense to me

**Comment by tannergooding** [comment]
How will the analyzer handle newline characters?

Meaning, if you have written `Text\nText`, but the actual source file contains `Text\r\nText`, will it complain?

**Comment by sharwell** [comment]
@tannergooding the analyzer treats `\n` and `\r\n` interchangeably. The code fix uses whatever line endings are already in the file.

**Comment by jmarolf** [comment]
alright, one question I have: How do I set enforcement levels for this so its an _error_ to not have a template header? I think I missed that somewhere. Are we saying that we'd have an option like `file_header_template_style` to configure enforcement?

```ini
[*.{cs,vb}]
file_header_template = <copyright file="{fileName}" company="Your Company Here">\nYour copyright text here\n</copyright>
file_header_template_style = prepend:error
file_header_template_style = replace:suggestion
```

where `prepend` and `replace` denote how the fixer should resolve the style violation?

**Comment by sharwell** [comment]
I don't think we'd need to configure the severity with an option when .editorconfig already lets you configure the severity by ID.

**Comment by jmarolf** [comment]
@sharwell How will the user know which ID to use? All of our other style options do not do this as far as I can tell.

**Comment by JoeRobich** [comment]
> where `prepend` and `replace` denote how the fixer should resolve the style violation?

Couldn't the correct fix be dependent on the file?

**Comment by jmarolf** [comment]
> Couldn't the correct fix be dependent on the file?

My assumption here was that the person applying this option would know if they place important non boiler-plate comments at the top of their files.

An example I've seen is something like this:
```csharp
/*
 * Implementation of this algorithm it taken form these papers:
 * ...
 */

using Microsoft.CodeAnalysis;

namespace Microsoft.CodeAnalysis {
	// Some set of classes
}
```

If I were to apply a fix to the file _automatically_ without user involvement for my entire project I should know if we already use copyright headers or not. If we do not then I likely want to prepend the new header and preserve whichever formatting already exists. If we do then I just want to replace the existing header with the new one and keep everything consistent. 

My assumption is that only the person applying the codefix knows which one is appropriate which is my bar for whether we should have and option.

**Comment by mavasani** [comment]
> @sharwell How will the user know which ID to use? All of our other style options do not do this as far as I can tell.

User does not need to know the ID. It should be possible to change it with the light bulb: https://docs.microsoft.com/en-us/visualstudio/code-quality/use-roslyn-analyzers?view=vs-2019#automatically-configure-rule-severity

**Comment by JoeRobich** [comment]
After talking offline to @jmarlof, I am on board.

 The point about how this fix would be automatically applied by Code Cleanup and dotnet-format as compelling. The `prepend` and `replace` option would only matter to automatic fixes. Users applying fixes through the IDE would have both methods available to them.

I would argue against tying this to a diagnostic id since we want this feature moved to the platform. Other CodeStyle options begin with `dotnet_` or `csharp_` but for this to be general I don't think those makes sense. I would suggest that the editorconfig key would be `file_header` and would default to `prepend:suggestion`.

**Comment by sharwell** [comment]
@jmarolf The only required code action is `replace`. The cases where `prepend` is needed are _extremely_ rare, and easily identified during review. When the `prepend` option is removed, the need to configure this diagnostic goes away, and we are back to the simple original proposal.

**Comment by vsnthdev** [comment]
How can I add a created date in the file header? 🤔

**Comment by jmarolf** [comment]
> How can I add a created date in the file header? 🤔

That is not supported currently. The idea is to have an identical file header across an entire repo not bespoke headers based on file creation time etc.
